src/main/java/org/openmaptiles/layers/WaterName.java

```diff
@@ -214,38 +218,63 @@ public void process(Tables.OsmMarinePoint element, FeatureCollector features) {
   @Override
   public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
     if (nullIfEmpty(element.name()) != null) {
-      Geometry centerlineGeometry = lakeCenterlines.get(element.source().id());
-      FeatureCollector.Feature feature;
-      int minzoom = 9;
-      String place = element.place();
-      String clazz;
-      if ("bay".equals(element.natural())) {
-        clazz = FieldValues.CLASS_BAY;
-      } else if ("sea".equals(place)) {
-        clazz = FieldValues.CLASS_SEA;
-      } else {
-        clazz = FieldValues.CLASS_LAKE;
-        minzoom = 3;
-      }
-      if (centerlineGeometry != null) {
-        // prefer lake centerline if it exists
-        feature = features.geometry(LAYER_NAME, centerlineGeometry)
-          .setMinPixelSizeBelowZoom(13, 6d * element.name().length());
-      } else {
-        // otherwise just use a label point inside the lake
-        feature = features.pointOnSurface(LAYER_NAME)
-          .setMinZoom(place != null && SEA_OR_OCEAN_PLACE.contains(place) ? 0 : 3)
-          .setMinPixelSize(128); // tiles are 256x256, so 128x128 is 1/4 of a tile
+      try {
+        Geometry centerlineGeometry = lakeCenterlines.get(element.source().id());
+        int minzoomCL = MINZOOM_BAY;
+        String place = element.place();
+        String clazz;
+        if ("bay".equals(element.natural())) {
+          clazz = FieldValues.CLASS_BAY;
+        } else if ("sea".equals(place)) {
+          clazz = FieldValues.CLASS_SEA;
+        } else {
+          clazz = FieldValues.CLASS_LAKE;
+          minzoomCL = MINZOOM_LAKE;
+        }
+        if (centerlineGeometry != null) {
+          // prefer lake centerline if it exists, but point will be also used if minzoom bellow 9 is calculated from area
+          // note: Here we're diverging from OpenMapTiles: For bays with minzoom (based on area) point is used between
+          // minzoom and Z8 and for Z9+ centerline is used, while OpenMaptiles sticks with points.
+          setupOsmWaterPolygonFeature(
+            element, features.geometry(LAYER_NAME, centerlineGeometry), clazz, minzoomCL)
+              .setMinPixelSizeBelowZoom(13, 6d * element.name().length());
+        }
+
+        // Show a label if a water feature covers at least 1/4 of a tile or z14+
+        Geometry geometry = element.source().worldGeometry();
+        double area = geometry.getArea();
+        int minzoom = (int) Math.floor(-1d - Math.log(Math.sqrt(area)) / LOG2);
+        if (place != null && SEA_OR_OCEAN_PLACE.contains(place)) {
+          minzoom = Math.clamp(minzoom, MINZOOM_SEA_AND_OCEAN, 14);
+        } else {
+          minzoom = Math.clamp(minzoom, MINZOOM_LAKE, 14);
```

Comment:
You should be able to replace this handling with a combination of min area and min zoom?

```java
feature.setMinPixelSizeBelowZoom(13, 128).setMinZoom(minzoom)
```

Reply:
Will that work if we're creating a point from that polygon with `features.pointOnSurface()`?

Reply:
Yes! I updated it to work that way in order to support use-cases like this here: https://github.com/onthegomap/planetiler/pull/720